woocommerce/wc-admin
# Add recommended payment methods

Fixes #6684 

This PR depends upon 9916-gh-Automattic/woocommerce.com  being merged.
The main change for this PR is adding a new card under the payments settings list with recommended payment providers, the list is retrieved from a woocommerce.com json, added in the PR mentioned above.

Additional changes:
- Add an embedded layout component that gets appended below the `wpBody` component, I added support for what I call a registry, allowing us to register react components for a specific page, see the [payment recommendations component](https://github.com/woocommerce/woocommerce-admin/blob/6f0e3c457469857b519b16d91f8f3476f69b1018/client/payments/index.ts#L18-L21) as an example. I added a custom history item that appends the `page`, `tab`, and/or `section` together with an `_`, allowing us to target the specific pages more easily by using the react router.
- Added a `useOptionsHydration` hook, which is mostly a wrapper of the `withOptionsHydration` function.
- Converted some (probably should finish) of the plugins data package to TS, and added support for the recommended payments query.

### Accessibility




### Screenshots

Recorded on my local dev, it loads quite a bit faster on jurassic:
![recommended-payments](https://user-images.githubusercontent.com/2240960/114200849-fa7be880-992b-11eb-9d0b-c08b39364add.gif)

### Detailed test instructions:
  
- If 9916-gh-Automattic/woocommerce.com is not merged yet (Merged now, so don't need to update this), replace the `$url` [here](https://github.com/woocommerce/woocommerce-admin/blob/6f0e3c457469857b519b16d91f8f3476f69b1018/src/PaymentPlugins.php#L50) with this: `https://gist.githubusercontent.com/louwie17/f5c0a1c57d3a9ead78fcf385782d8a05/raw/20809ef91680f86d73f47caf6349055512ec695f/recommended-payments.json`, note that the icons won't show unless you also serve up the WooCommerce.com PR mentioned earlier. And start your local environment or generate a `test:zip` for jurassic
- Make sure your store location is filled out and within US | PR | AU | CA | GB | IE | NZ and enable debug messages in console using `localStorage.setItem( 'debug', 'wc-admin:*' );`
- Visit **Woocommerce > Settings > Payments** you might have to wait a couple seconds, but it should show a card with **Recommended ways to get paid** listing 3 different payment providers (this data comes from the json referred to earlier).
- It should display a `wcadmin_settings_payments_recommendations_pageview` track
- Click `Get started` on one of the providers, it will show a loading icon (installing the plugin), once done it should redirect you to the plugin set up page.
- It should trigger a `wcadmin_settings_payments_recommendations_setup` track with an `extension_selected` attribute set to the extension slug.
- Go back to the payment settings page
- Notice how the plugin you had previously installed and activated does not display anymore.
- Click on the 3 dots of the card, click `Hide this`, it should make the card disappear, and also not be displayed on refresh. Deleting `woocommerce_show_marketplace_suggestions` option, should make it appear again.
Hiding the card should trigger a `wcadmin_settings_payments_recommendations_dismiss` track.

After making sure `woocommerce_show_marketplace_suggestions` is deleted and the card shows up again.
- Go to **WooCommerce > Settings > Advanced > WooCommerce.com** and un-select **Show Suggestions** and save
- Go to the payments setting screen again, the card should not be displayed, nor should any requests have gone out to recommended-payments (it should get the component chunk though).

Enable the **Show Suggestions** in the **Woocommerce.com** settings again, and change the store's country in **WooCommerce > Settings > General** to something not in this list: US | PR | AU | CA | GB | IE | NZ. (You can pick France). 
- Go back to the Payment settings screen, the recommended card should not show up.
